onnxruntime/core/optimizer/constant_sharing.cc

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <variant>
+
+#include "core/framework/tensorprotoutils.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/constant_sharing.h"
+#include "core/optimizer/utils.h"
+
+namespace onnxruntime {
+
+namespace {
+
+// Supports limited data types.
+static constexpr std::array supported_data_types{
+    ONNX_NAMESPACE::TensorProto_DataType_FLOAT,
+    ONNX_NAMESPACE::TensorProto_DataType_FLOAT16,
+    ONNX_NAMESPACE::TensorProto_DataType_DOUBLE,
+    ONNX_NAMESPACE::TensorProto_DataType_INT32,
+    ONNX_NAMESPACE::TensorProto_DataType_INT64,
+};
+
+bool IsSupportedDataType(int32_t data_type) {
+  return std::find(std::begin(supported_data_types), std::end(supported_data_types), data_type) !=
+         std::end(supported_data_types);
```

Comment:
std::array has begin() and end() methods doesn't it (or maybe cbegin/cend are needed)?